rejistry++/src/RegistryKey.cpp

```diff
@@ -68,16 +68,29 @@ namespace Rejistry {
 
     RegistryKey::RegistryKeyPtrList RegistryKey::getSubkeyList() const {
         std::vector<RegistryKey *> subkeys;
-        NKRecord::NKRecordPtrList nkRecordList = _nk->getSubkeyList()->getSubkeys();
+        SubkeyListRecord::SubkeyListRecordPtr subkeyListRecordPtr = _nk->getSubkeyList();
+        NKRecord::NKRecordPtrList nkRecordList = subkeyListRecordPtr->getSubkeys();
         NKRecord::NKRecordPtrList::iterator it;
         for (it = nkRecordList.begin(); it != nkRecordList.end(); ++it) {
             subkeys.push_back(new RegistryKey(*it));
         }
+        delete subkeyListRecordPtr;
```

Comment:
I don't believe getSubkeyList() made a new copy of the data, so it probably shouldn't be freed here. This also applies to the four other changes to this file.

Reply:
Yes they do.
Line 71 SubkeyListRecord::SubkeyListRecordPtr subkeyListRecordPtr = _nk->getSubkeyList();
getSubkeyList in NKRecord.cpp line 151
``
            return new EmptySubkeyList(_buf, 0);
``
or line 161
``
    return c->getSubkeyList();
``
getSubkeyList() in Cell.cpp line 101
``
            return getLHRecord();
``
which in Cell.cpp line 71, returns a new LHRecord.
``
    SubkeyListRecord::SubkeyListRecordPtr Cell::getLHRecord() const {
        return new LHRecord(_buf, getAbsoluteOffset(DATA_OFFSET));
    }
``

I have the vld stacktrace to show me the leaks.
> 

---------- Block 4084816 at 0x000000003594C940: 32 bytes ----------
  Leak Hash: 0x082E05D6, Count: 60, Total 1920 bytes
  Call Stack:
    ntdll.dll!RtlAllocateHeap()
    f:\dd\vctools\crt\vcstartup\src\heap\new_scalar.cpp (19): tsk_logical_imager.exe!operator new() + 0xA bytes
    c:\users\jkho\git\sleuthkit\rejistry++\src\cell.cpp (71): tsk_logical_imager.exe!Rejistry::Cell::getLHRecord() + 0xA bytes
    c:\users\jkho\git\sleuthkit\rejistry++\src\cell.cpp (101): tsk_logical_imager.exe!Rejistry::Cell::getSubkeyList() + 0xC bytes
    c:\users\jkho\git\sleuthkit\rejistry++\src\nkrecord.cpp (163): tsk_logical_imager.exe!Rejistry::NKRecord::getSubkeyList() + 0x11 bytes
    c:\users\jkho\git\sleuthkit\rejistry++\src\registrykey.cpp (80): tsk_logical_imager.exe!Rejistry::RegistryKey::getSubkey() + 0x2B bytes
    c:\users\jkho\git\sleuthkit\tools\logicalimager\regparser.cpp (405): tsk_logical_imager.exe!RegParser::findKey() + 0x2A bytes
    c:\users\jkho\git\sleuthkit\tools\logicalimager\regparser.cpp (121): tsk_logical_imager.exe!RegParser::getKey() + 0x12 bytes
    c:\users\jkho\git\sleuthkit\tools\logicalimager\registryanalyzer.cpp (214): tsk_logical_imager.exe!RegistryAnalyzer::analyzeSAMUsers() + 0x26 bytes
    c:\users\jkho\git\sleuthkit\tools\logicalimager\tsk_logical_imager.cpp (579): tsk_logical_imager.exe!main()
    f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl (65): tsk_logical_imager.exe!invoke_main()
    f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl (253): tsk_logical_imager.exe!__scrt_common_main_seh() + 0x5 bytes
    f:\dd\vctools\crt\vcstartup\src\startup\exe_common.inl (296): tsk_logical_imager.exe!__scrt_common_main()
    f:\dd\vctools\crt\vcstartup\src\startup\exe_main.cpp (17): tsk_logical_imager.exe!mainCRTStartup()
    KERNEL32.DLL!BaseThreadInitThunk() + 0x14 bytes
    ntdll.dll!RtlUserThreadStart() + 0x21 bytes
  Data:
    A0 FB AF C3    F6 7F 00 00    60 14 8D 35    B8 01 00 00     ........ `..5....
    D4 2F 00 00    CD CD CD CD    08 00 00 00    CD CD CD CD     ./...... ........
>